### PR TITLE
Rational number limit

### LINKED ISSUE #3327: SOL-008 CPU blow-up using huge bignums literals
**Description**

The Solidity compiler accepts numeric literals of arbitrary precision. But computing those literals can take arbitrary high time.
When the following sample contract is compiled, the compiler runs for several hours (we’ve not waited until it finishes):

contract BIGNUMTEST {
function bignum() {
uint c;
c=1E10000000;
}
}

**Recommendations**

Limit the size of numeric literals.

### LINKED ISSUE #3759: Constant folding with arbitrary precision arithmetic?
contract c {
function f() returns (int) {
return 4 ** 4 ** 2 ** 4 ** 4 ** 4 ** 4 ** 4 ** 4 ** 4 ** 4 ** 4 ** 4 * 2;

}

Causes the compiler to hang (surprisingly, not exceed memory limit) for hours.  Not sure a "bug," but worth reporting as sub-optimal.

### PR DESCRIPTION
**Pull request by erak**
Fixes #3327
Fixes #3759

### CODE REVIEW

File: libsolidity/ast/Types.cpp [hunk 1]
```diff
@@ -905,6 +916,11 @@ TypePointer RationalNumberType::binaryOperatorResult(Token::Value _operator, Typ
 			uint32_t exponent = abs(other.m_value).numerator().convert_to<uint32_t>();
 			bigint numerator = pow(m_value.numerator(), exponent);
 			bigint denominator = pow(m_value.denominator(), exponent);
+
+			// Limit size to 4096 bits
+			if (numerator >= pow(bigint(2), rationalMaxBits) - 1)
```
Comment (axic) [inline libsolidity/ast/Types.cpp]: This `pow` should be evaluated in the const variable. Also you can use shift instead of pow for power-of-2.

Comment (axic) [reply libsolidity/ast/Types.cpp]: I don't think it is enough to check for the value **after** it has ben evaluated. It should be guessed before evaluating with `pow`.

One of the main problems here is the DoS vector (which is just annoying for a user, for example running the compiler in the browser).

Comment (axic) [reply libsolidity/ast/Types.cpp]: I don't think it is enough to check for the value **after** it has ben evaluated. It should be guessed before evaluating with `pow`.

One of the main problems here is the DoS vector (which is just annoying for a user, for example running the compiler in the browser).

**Review by axic — CHANGES_REQUESTED** [review]

File: libsolidity/ast/Types.cpp (hunk 1 above)
Comment (bit-shift) [reply libsolidity/ast/Types.cpp]: I've started testing with two examples of those DoS cases and they're not hanging anymore. After the limit is reached, the function is still called, but  ``m_value.denominator()`` is constant from there on and calling ``pow()`` results in the same large number. 
Currently the error messages don't help to understand the issue I guess, but there might be some options for makes those richer.

**Comment by erak** [comment]
@axic I've started testing with two examples of those DoS cases and they're not hanging anymore. After the limit is reached, the function is still called, but ``m_value.denominator()`` is constant from there on and calling ``pow()`` results in the same large number.
Currently the error messages don't help to understand the issue I guess, but there might be some options for making them richer.

File: libsolidity/ast/Types.cpp [hunk 2]
```diff
@@ -43,6 +43,14 @@ using namespace std;
 using namespace dev;
 using namespace dev::solidity;
 
+namespace
+{
+
+uint32_t static const exponentLimit = 9999;
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: We don't really use `uint32_t` explicitly in the rest of the codebase. I think I almost exclusively use `size_t`.

File: libsolidity/ast/Types.cpp [hunk 3]
```diff
@@ -447,7 +455,7 @@ bigint IntegerType::maxValue() const
 }
 
 TypePointer IntegerType::binaryOperatorResult(Token::Value _operator, TypePointer const& _other) const
-{
+{
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Surplus whitespace.

File: libsolidity/ast/Types.cpp [hunk 4]
```diff
@@ -677,12 +685,15 @@ tuple<bool, rational> RationalNumberType::isValidLiteral(Literal const& _literal
 		}
 		else if (expPoint != _literal.value().end())
 		{
-			// parse the exponent
+			// parse the exponent and check limits
 			bigint exp = bigint(string(expPoint + 1, _literal.value().end()));
 
 			if (exp > numeric_limits<int32_t>::max() || exp < numeric_limits<int32_t>::min())
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: This can now be removed.

File: libsolidity/ast/Types.cpp [hunk 5]
```diff
@@ -677,12 +685,15 @@ tuple<bool, rational> RationalNumberType::isValidLiteral(Literal const& _literal
...
 				return make_tuple(false, rational(0));
 
+			if (abs(exp) >= exponentLimit)
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Perhaps it would be better to re-use the same helper in all functions that compute exp for rational numbers. Just pass the base and the exponent, and it will determine if the resulting number still fits our precision.

Comment (chriseth) [reply libsolidity/ast/Types.cpp]: Also, if we know the base, we can better estimate the size of the resulting number.

**Review by chriseth — CHANGES_REQUESTED** [review]

**Comment by chriseth** [comment]
Please also add a test for bit-shifts.

File: libsolidity/ast/Types.cpp (hunk 2 above)
Comment (axic) [reply libsolidity/ast/Types.cpp]: `size_t` or `unsigned` is most common in the codebase

**Comment by axic** [comment]
> After the limit is reached, the function is still called, but m_value.denominator() is constant from there on and calling pow() results in the same large number.

For those test cases it probably is the case, however for the case of doing a very large exp, it will just "hang" or take a lot of time.

Please add such a test case too.

**Comment by axic** [comment]
Maybe useful to look at the test cases of #3348 (an earlier attempt of the same problem).

Additionally, some random notes:
- `uint a = 1E1000000 ** 1E1000000;`
- `100000000000E1250 = 1E1261`

File: test/libsolidity/syntaxTests/types/rational_number_literal_limit.sol [hunk 6]
```diff
@@ -0,0 +1,10 @@
+contract c {
+    function bignum() public {
+        uint a;
+        uint b;
+        a = 1e9999;
+				b = 13456232453246423445233516816351681345623245324642344523351681635168.7161354161363516163161354168154E1200;
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/types/rational_number_literal_limit.sol]: Indentation

File: libsolidity/ast/Types.cpp [hunk 7]
```diff
@@ -40,10 +40,35 @@
 
 #include <limits>
```
Comment (ekpyron) [inline libsolidity/ast/Types.cpp]: I think STL includes usually go after boost includes (I think the general schema is: most specific first, most generic last).

**Review by ekpyron — CHANGES_REQUESTED** [review]

File: libsolidity/ast/Types.cpp [hunk 8]
```diff
@@ -40,10 +40,35 @@
...
+	using boost::multiprecision::log10;
+	using boost::multiprecision::ceil;
+	using bigfloat = boost::multiprecision::number<boost::multiprecision::cpp_dec_float<50>>;
+
+	size_t const bitsMax = 4096;
+	bigint const baseMax = (bigint(2) << bitsMax) - 1;
+	if(_base > baseMax)
+		return false;
+
+	bigfloat bitsNeeded = bigfloat(_exp) * ceil(log10(bigfloat(_base + 1)));
```
Comment (ekpyron) [inline libsolidity/ast/Types.cpp]: Maybe ``log2`` instead of ``log10``?

File: libsolidity/ast/Types.cpp [hunk 9]
```diff
@@ -40,10 +40,35 @@
...
 using namespace dev::solidity;
 
+namespace
+{
+
+bool fitsPrecision(bigint const _base, bigint const _exp)
+{
+	using boost::multiprecision::log10;
+	using boost::multiprecision::ceil;
+	using bigfloat = boost::multiprecision::number<boost::multiprecision::cpp_dec_float<50>>;
```
Comment (axic) [inline libsolidity/ast/Types.cpp]: Does that mean it allows up to 50 decimal digits? What happens if `_base + 1` is larger than 50 decimal digits?

According to boost these are also defined:
```
typedef number<cpp_dec_float<50> > cpp_dec_float_50;
typedef number<cpp_dec_float<100> > cpp_dec_float_100;
```

Comment (erak) [reply libsolidity/ast/Types.cpp]: @axic The precision will be reduced after creating ``cpp_dec_float_50`` with ``base + 1``. That means a if ``base`` has more than 50 digits, any additional digit will be represented with a zero. This does not affect the size of it.

Comment (axic) [reply libsolidity/ast/Types.cpp]: Can you add a test for this? Boost had some bugs where an overflow (with shifts) caused the entire value to be 0.

File: libsolidity/ast/Types.cpp (hunk 8 above)
Comment (erak) [reply libsolidity/ast/Types.cpp]: That was a defect ;) ``log10`` is what I wanted but this formula calculated the number of digits and not the number of bits. I've added another formula which estimates the minimum number of bits needed.

File: libsolidity/ast/Types.cpp (hunk 7 above)
Comment (bit-shift) [reply libsolidity/ast/Types.cpp]: Done.

File: libsolidity/ast/Types.cpp [hunk 10]
```diff
@@ -37,13 +37,39 @@
...
+	using boost::multiprecision::ceil;
+	using bigfloat = boost::multiprecision::cpp_dec_float_50;
+
+	size_t const bitsMax = 4096;
+	bigint const baseMax = (bigint(2) << bitsMax) - 1;
+	if(_base > baseMax)
+		return false;
+
+	bigfloat digitsNeeded = bigfloat(_exp) * ceil(log10(bigfloat(_base + 1)));
+	bigfloat bitsNeeded = ceil((digitsNeeded - 1) * log2(bigfloat(10)));
```
Comment (ekpyron) [inline libsolidity/ast/Types.cpp]: Why calculate ``digitsNeeded`` as an intermediate result?
You might as well just use ``log2`` directly to get ``bitsNeeded`` instead of multiplying with ``log2(10)`` later...

``log2(10) * log10(a) = log2(a)``

I don't think using 10 as base has any advantage for choosing the correct way of rounding things...

What you want to have in the end is: ``2 ^ 4096 - 1 >= _base^_exp`` or, since it's integers:
``2^4096 > _base^_exp``.

Hence you want ``log2(2^4096) > log2(_base^_exp)``.
Hence `` 4096 > _exp * log2(_base)``.

You only need to compensate for potential rounding error in ``log2(_base)``.

**Review by ekpyron — DISMISSED** [review]

File: libsolidity/ast/Types.cpp (hunk 10 above)
Comment (ekpyron) [reply libsolidity/ast/Types.cpp]: Safest thing would probably be ``bitsNeeded = _exp * (floor(log2(_base)) + 1)``.

Comment (ekpyron) [reply libsolidity/ast/Types.cpp]: Although you may want to treat ``_base <= 1`` separately :-).

Comment (ekpyron) [reply libsolidity/ast/Types.cpp]: Also: do you only use ``fitsPrecision`` for ``_base >= 0``? You'd probably need to treat negative ``_base`` differently...
And probably also add tests for that.

**Review by axic — CHANGES_REQUESTED** [review]
Also needs a test for a larger base with a smaller exponent, which still exceeds the limit.

File: libsolidity/ast/Types.cpp [hunk 11]
```diff
@@ -37,13 +37,39 @@
...
+bool fitsPrecision(bigint const _base, bigint const _exp)
+{
+	using boost::multiprecision::log10;
+	using boost::multiprecision::log2;
+	using boost::multiprecision::ceil;
+	using bigfloat = boost::multiprecision::cpp_dec_float_50;
+
+	size_t const bitsMax = 4096;
+	bigint const baseMax = (bigint(2) << bitsMax) - 1;
+	if(_base > baseMax)
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Space after `if`

File: libsolidity/ast/Types.cpp [hunk 12]
```diff
@@ -37,13 +37,39 @@
...
 #include <limits>
 
 using namespace std;
 using namespace dev;
 using namespace dev::solidity;
 
+namespace
+{
+
+bool fitsPrecision(bigint const _base, bigint const _exp)
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Why not `const&`?

**Review by chriseth — CHANGES_REQUESTED** [review]

File: libsolidity/ast/Types.cpp [hunk 13]
```diff
@@ -949,7 +1042,10 @@ TypePointer RationalNumberType::binaryOperatorResult(Token::Value _operator, Typ
 			else if (other.m_value > numeric_limits<uint32_t>::max())
 				return TypePointer();
 			uint32_t exponent = other.m_value.numerator().convert_to<uint32_t>();
-			value = rational(m_value.numerator() / pow(bigint(2), exponent), 1);
+			if (exponent > msb(m_value.numerator()))
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: `msb` is an extremely short name to be used without namespace. Is that imported somewhere?

File: libsolidity/ast/Types.cpp [hunk 14]
```diff
@@ -689,31 +754,33 @@ tuple<bool, rational> RationalNumberType::isValidLiteral(Literal const& _literal
...
 			if (!get<0>(base))
 				return make_tuple(false, rational(0));
 			value = get<1>(base);
 
 			if (exp < 0)
 			{
 				exp *= -1;
+				if (!fitsPrecisionBase10(abs(value.denominator()), exp))
+					return make_tuple(false, rational(0));
 				value /= boost::multiprecision::pow(
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: add assert (or check) that `exp > numeric_limits<int32_t>::max() || exp < numeric_limits<int32_t>::min()`

File: libsolidity/ast/Types.cpp [hunk 15]
```diff
@@ -37,13 +37,78 @@
...
+		return true;
+	if (mostSignificantBaseBit > bitsMax) // _base >= 2 ^ 4096
+		return false;
+
+	bigint bitsNeeded = _exp * (mostSignificantBaseBit + 1);
+
+	return bitsNeeded <= bitsMax;
+}
+
+bool fitsPrecisionBaseX(
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Please add a comment about the rough semantics of the function.

File: libsolidity/ast/Types.cpp [hunk 16]
```diff
@@ -37,13 +37,78 @@
...
+	solAssert(_expBase10 >= 0, "");
+
+	if (!_mantissa)
+		return true;
+
+	// _mantissa > 0
+
+	size_t const bitsMax = 4096;
+
+	auto mostSignificantBit = msb(_mantissa);
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Please add an explicit type.

File: libsolidity/ast/Types.cpp (hunk 13 above)
Comment (ekpyron) [reply libsolidity/ast/Types.cpp]: I don't think this will be a problem - ``msb`` is not imported in the namespace anyways, but only found via argument-dependent lookup and, if there will be a disambiguity at some point, this should raise an error and can be adjusted then.

File: libsolidity/ast/Types.cpp [hunk 17]
```diff
@@ -44,6 +44,71 @@ using namespace std;
 using namespace dev;
 using namespace dev::solidity;
 
+namespace
+{
+
+// checks whether (_base ^ _exp) fits into 4096 bits
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: I think `**` would be better here. Also we generally use `///` ("javadoc") to document functions.

File: libsolidity/ast/Types.cpp [hunk 18]
```diff
@@ -44,6 +44,71 @@ using namespace std;
...
+bool fitsPrecisionExp(bigint const& _base, bigint const& _exp)
+{
+	solAssert(_base >= 0, "");
+
+	if (!_base)
+		return true;
+
+	// _base > 0
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Why not turn this into an assert and remove the one above?

File: libsolidity/ast/Types.cpp [hunk 19]
```diff
@@ -44,6 +44,71 @@ using namespace std;
...
+
+	size_t const bitsMax = 4096;
+
+	auto mostSignificantBaseBit = msb(_base);
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Please use an explicit type here.

**Review by chriseth — CHANGES_REQUESTED** [review]

**Comment by ekpyron** [comment]
OK - ``boost::multiprecision::msb`` is only introduced with boost 1.55 and trusty is still on boost 1.54 (somehow I thought it'd be at least 1.57) :-D.

I'll try to find a good workaround over the weekend.

The ``ci/circleci: test_x86`` result ``Too long with no output (exceeded 10m0s)`` is not a good sign as well, though...

**Comment by chriseth** [comment]
@ekpyron were you able to find something? Otherwise I will pick this up later.

**Comment by ekpyron** [comment]
@chriseth I was just about to look into it. Unfortunately I think we may have to use ``floor(log2(...))`` instead of ``msb`` again and therefore also deal with some ``bigfloat`` type... I'll check whether there's a better option.

**Comment by ekpyron** [comment]
@chriseth This now contains a workaround for boost < 1.55, which is a naive (but therefore sufficiently simple) implementation of ``msb``. We have to keep an eye on the running times of the tests, though.

**Comment by ekpyron** [comment]
Now the ``ci/circleci: test_x86`` test works again as well and doesn't seem to take longer than before the PR - not sure what the problem was before...

**Review by chriseth — APPROVED** [review]
